aten/src/ATen/FunctionalInverses.cpp

```diff
@@ -0,0 +1,191 @@
+
+#include <ATen/FunctionalInverses.h>
+
+#include <ATen/ATen.h>
+#include <ATen/ExpandUtils.h>
+namespace at {
+namespace functionalization {
+namespace impl {
+
+// This logic is similar to autograd code for view backwards calls.
+// We can't easily share it though, because (eventually) these functions
+// will all call `permute/unsqueeze_copy()` instead of `permute/unsqueeze`.
+
+Tensor permute_inverse(const Tensor& self, IntArrayRef dims) {
+  // invert the permutation
+  auto ndims = dims.size();
+  std::vector<int64_t> dims_(ndims);
+  for(const auto i : c10::irange(ndims)) {
+    dims_[at::maybe_wrap_dim(dims[i], ndims)] = i;
+  }
+  return self.permute(dims_);
+}
+
+Tensor unsqueeze_to(const Tensor & self, IntArrayRef sizes) {
+  auto result = self;
+
+  int64_t nDims = sizes.size();
+  for(const auto dim : c10::irange(nDims)) {
+    if (sizes[dim] == 1) {
+      result = result.unsqueeze(dim);
+    }
```

Comment:
Should there be something like, asserting that the sizes match if no unsqueeze happens?

Reply:
I don't think that can happen from user error, right? Just if there's a bug in the pass somewhere. But some defensive asserts seem reasonable